### ISSUE TITLE
Whether there exists the [length] token?

### ISSUE THREAD

**Issue by yfzhang114** [facebookresearch/Mask-Predict#17]
I have a question about the length token. I cannot find the length token in the input source sentences
![image](https://user-images.githubusercontent.com/53161950/121762607-8fd46c80-cb69-11eb-9aaa-bbfe3b409ccf.png)

**Comment by hjx999222** [comment]
Hello, I have the same problem now. Have you solved yet?

**Comment by yair-schiff** [comment]
I am also digging into this codebase. My current understanding is that the `length` token isn't part of the src input sequence but gets created during the encoder `forward` pass [here](https://github.com/facebookresearch/Mask-Predict/blob/11c753526ebbbf5b946d1305a613b830f4f29232/fairseq/models/bert_seq2seq.py#L542-L543).
```python
len_tokens = self.embed_lengths(src_tokens.new(src_tokens.size(0), 1).fill_(0))
x = torch.cat([len_tokens, x], dim=1)
```

Predictions are generated later on in this method [here](https://github.com/facebookresearch/Mask-Predict/blob/11c753526ebbbf5b946d1305a613b830f4f29232/fairseq/models/bert_seq2seq.py#L560-L562).
```python
predicted_lengths_logits = torch.matmul(x[0, :, :], self.embed_lengths.weight.transpose(0, 1)).float()
predicted_lengths_logits[:, 0] += float('-inf')   # Cannot predict the len_token
predicted_lengths = F.log_softmax(predicted_lengths_logits, dim=-1)
```

(Note that the reason, the first token (which is the appended, newly created `length` token is indexed like this `x[0, :, :]` is because the sequence gets transposed at this [line](https://github.com/facebookresearch/Mask-Predict/blob/11c753526ebbbf5b946d1305a613b830f4f29232/fairseq/models/bert_seq2seq.py#LL547C9-L547C30):  `x = x.transpose(0, 1)`)
